### PR TITLE
Change default type name handling to None

### PR DESCRIPTION
**Pull request by jpalac**
* Add new serialiser NewtonsoftJsonSerializer
* Mark NewtonsoftSerializer as obsolete - to error in this version and be removed in next
* This will be applied to version 2.3 as well which will result in the warning
* Adjust existing tests to work with new serialiser
* Add new tests to ensure embedded $types do not work with TypeNameHandling default value of None

### CODE REVIEW

File: src/NServiceBus.Newtonsoft.Json/NewtonsoftSerializer.cs [hunk 1]
```diff
@@ -22,9 +26,8 @@ public class NewtonsoftSerializer : SerializationDefinition
                 var writerCreator = settings.GetWriterCreator();
                 var serializerSettings = settings.GetSettings();
                 var contentTypeKey = settings.GetContentTypeKey();
-                return new JsonMessageSerializer(mapper, readerCreator, writerCreator, serializerSettings, contentTypeKey);
+                return new JsonMessageSerializer(mapper, readerCreator, writerCreator, serializerSettings, contentTypeKey, global::Newtonsoft.Json.TypeNameHandling.Auto);
```
Comment (mikeminutillo) [inline src/NServiceBus.Newtonsoft.Json/NewtonsoftSerializer.cs:29]: Once we have hit the version of `TreatAsErrorFromVersion` we can replace the implementation with `throw new NotImplementedException();`

We might want to cherry pick this change back to the `release-2.2` branch first because we'll need the implementation there.

File: src/NServiceBus.Newtonsoft.Json/NewtonsoftConfigurationExtensions.cs [hunk 2]
```diff
@@ -80,6 +116,22 @@ public static void ContentTypeKey(this SerializationExtensions<NewtonsoftSeriali
...
+        /// <summary>
+        /// Configures string to use for <see cref="Headers.ContentType"/> headers.
+        /// </summary>
+        /// <remarks>
+        /// Defaults to <see cref="ContentTypes.Json"/>.
+        /// This setting is required when this serializer needs to co-exist with other json serializers.
+        /// </remarks>
+        /// <param name="config">The <see cref="SerializationExtensions{T}"/> instance.</param>
+        /// <param name="contentTypeKey">The content type key to use.</param>
+        public static void ContentTypeKey(this SerializationExtensions<NewtonsoftJsonSerializer> config, string contentTypeKey)
```
Comment (mikeminutillo) [inline src/NServiceBus.Newtonsoft.Json/NewtonsoftConfigurationExtensions.cs:128]: Note that all of the overloads that target the original type can be deprecated as well, and they won't need an implementation on master. They can just throw `NotImplementedException`

In some ways, adding the deprecation attribute will just create noise. Imagine the following scenario:

```csharp
var serializer = endpointConfiguration.UseSerializer<NewtonsoftSerializer>();
serializer.ContentTypeKey("Blah");
```

If we deprecate the `ContentTypeKey()` configuration method as well, the user will get two compile time errors, even though they only need to change the first line.

We could maybe resolve this with a marker interface. If we created an interface like `INewtonsoftJsonSerializer` and put it on both serializer definitions then these method signatures could become something like:

```csharp
public static void ContentTypeKey<T>(this SerializationExtensions<T> config, string contentTypeKey)
  where T : INewtonsoftJsonSerializer
```

It does mean adding the marker interface to the public API surface and we'll have to remove that at some point too.

File: src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs [hunk 3]
```diff
@@ -23,14 +23,23 @@ class JsonMessageSerializer : IMessageSerializer
...
+                settings = new JsonSerializerSettings { TypeNameHandling = typeNameHandling };
+                if (typeNameHandling == TypeNameHandling.Auto)
+                {
+                    log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json use TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<NewtonsoftSerializer>().Settings'. Refer to the Json.NET serializer documentation at https://docs.particular.net/ for further details.");
+                }
+            }
+
+            if (settings.TypeNameHandling == TypeNameHandling.Auto && typeNameHandling != TypeNameHandling.Auto)
+            {
+                log.Warn($"Use of TypeNameHandling.Auto is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible");
```
Comment (mikeminutillo) [inline src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:42]: I think if a user is using the old definition and they don't provide settings they will see both of these log warnings.

Line 33 will set to Auto
Line 34 will be true and the log on line 36 will be output
Line 40 is also true and the log on line 42 will be output

Comment (jpalac) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:42]: The old definition will have the default passed in as auto, hence the second warning will not be shown as it excludes that.

Comment (mikeminutillo) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:42]: You are right! My bad.

File: src/NServiceBus.Newtonsoft.Json/NewtonsoftSerializer.cs (hunk 1 above)
Comment (jpalac) [reply src/NServiceBus.Newtonsoft.Json/NewtonsoftSerializer.cs]: I have replaced all implementations with `NotImplementedException();`
We will manually add the relevant code to 2.3 instead of cherry picking.

File: src/NServiceBus.Newtonsoft.Json/NewtonsoftConfigurationExtensions.cs (hunk 2 above)
Comment (jpalac) [reply src/NServiceBus.Newtonsoft.Json/NewtonsoftConfigurationExtensions.cs:77]: I have moved all obsolete methods into their own file and have marked the methods as obsolete to be in line with how we do things in core.

**Review by mikeminutillo — APPROVED** [review]

File: src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs [hunk 4]
```diff
@@ -23,14 +23,23 @@ class JsonMessageSerializer : IMessageSerializer
...
+                settings = new JsonSerializerSettings { TypeNameHandling = typeNameHandling };
+                if (typeNameHandling == TypeNameHandling.Auto)
+                {
+                    log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json use TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<NewtonsoftSerializer>().Settings'. Refer to the Json.NET serializer documentation at https://docs.particular.net/ for further details.");
```
Comment (ramonsmits) [inline src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:36]: A quick review I was wondering how a user could even set this to Auto without passing `JsonSerializerSettings`? It seems this is only internal? If its not internal... shouldn't it be internal?

Comment (ramonsmits) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs:36]: Second, I think we could modify the url to direct to the docs search page with the component ID.

Comment (jpalac) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: You are correct - in fact the typeNameHandling parameter is not needed at all in V3 as the old serializer is not implemented and throws an error. 
We have removed the parameter and the warning message which no longer applies.

**Review by ramonsmits — APPROVED** [review]
